controllers/submission.go

```diff
@@ -123,14 +126,148 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
+	}
+
+	var submissions model.Submissions
+	var keys []*datastore.Key
+	keys, err = model.NewQueryForSubmission().
+		Ancestor(resultKey).
+		Filter("Task =", taskKey).
+		Order("Time").
+		GetAll(ctx, &submissions)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	json.NewEncoder(w).Encode(submissions.Key(keys))
+	return http.StatusOK, nil
+}
+
+func GetSubmissionByKey(ctx context.Context, w http.ResponseWriter, r *http.Request) (status int, err error) {
+	if r.Method != "GET" {
+		return http.StatusMethodNotAllowed, nil
+	}
+
+	// TODO(victorbalan): Check if user is company or if user is parent of result
+	// else return http.StatusUnauthorized
+	_, ok := passenger.FromContext(ctx)
+	if !ok {
+		return http.StatusUnauthorized, nil
+	}
+
+	submissionKey, err := datastore.DecodeKey(mux.Vars(r)["key"])
+	if err != nil {
+		return http.StatusNotFound, err
+	}
+
+	var submission model.Submission
+	if err = datastore.Get(ctx, submissionKey, &submission); err != nil {
+		return
+	}
+
+	codeFilesURLs, err := util.ExposeMultiURL(ctx, submission.Code.Bucket, submission.Code.Name)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	s := struct {
+		Time     time.Time
+		CodeURLs []string
+		Language string
+	}{
+		Time:     submission.Time,
+		Language: submission.Language,
+		CodeURLs: codeFilesURLs,
+	}
+
+	json.NewEncoder(w).Encode(s)
+	return http.StatusOK, nil
+}
+
+func GetTestResultsForSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request) (status int, err error) {
+	if r.Method != "GET" {
+		return http.StatusMethodNotAllowed, nil
+	}
+
+	// TODO(victorbalan): Check if user is company or if user is parent of result
+	// else return http.StatusUnauthorized
+	_, ok := passenger.FromContext(ctx)
+	if !ok {
+		return http.StatusUnauthorized, nil
+	}
+
+	submissionKey, err := datastore.DecodeKey(mux.Vars(r)["key"])
+	if err != nil {
+		return http.StatusNotFound, err
+	}
+
+	keys, err := datastore.NewQuery("").
+		Ancestor(submissionKey).
+		Filter("__key__ >", submissionKey).
+		KeysOnly().
+		GetAll(ctx, nil)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if len(keys) == 0 {
+		json.NewEncoder(w).Encode([]string{})
+		return http.StatusOK, nil
+	}
+
+	switch keys[0].Kind() {
+	case model.JunitTestResultKind:
+		var results model.JunitTestResults
+		_, err = datastore.NewQuery(keys[0].Kind()).
+			Ancestor(submissionKey).
+			GetAll(ctx, &results)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+		json.NewEncoder(w).Encode(results)
+		return http.StatusOK, nil
+	case model.DiffTestResultKind:
+		var results model.DiffTestResults
+		_, err = datastore.NewQuery(keys[0].Kind()).
+			Ancestor(submissionKey).
+			GetAll(ctx, &results)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+		json.NewEncoder(w).Encode(results)
+		return http.StatusOK, nil
+	default:
+		w.Write([]byte("[]"))
+		return http.StatusOK, nil
```

Comment:
You return `return http.StatusOK, nil` in all cases, so move it after the switch.